x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/pivot/DateHistogramGroupSource.java

```diff
@@ -248,6 +256,71 @@ public static DateHistogramGroupSource fromXContent(final XContentParser parser,
         return lenient ? LENIENT_PARSER.apply(parser, null) : STRICT_PARSER.apply(parser, null);
     }
 
+    public static boolean isInvalidFormat(String format, DateHistogramInterval interval) {
+        if (format == null || interval == null) {
+            return false;
+        }
+        DateFormatter formatter = DateFormatter.forPattern(format);
+        String expression = interval.toString();
+        String epochFormat = formatter.formatMillis(Instant.EPOCH.toEpochMilli());
+        String epochPlusIntervalFormat = formatter.formatMillis(Instant.EPOCH.toEpochMilli() + interval.estimateMillis());
+        if (DateHistogramAggregationBuilder.DATE_FIELD_UNITS.containsKey(expression)) {
+            // The estimated rounding defaults to 30 days for monthly, EPOCH starts in January, which has 31 days.
+            if (Rounding.DateTimeUnit.MONTH_OF_YEAR.equals(DateHistogramAggregationBuilder.DATE_FIELD_UNITS.get(expression))) {
```

Comment:
This is a horrible breach of encapsulation.

I wonder if you could make the code clearer and less verbose as well as not relying on an internal detail of `DateHistogramInterval.estimateMillis()` by using the Java 8 classes more directly?

```
    String expression = interval.toString();
	String epochFormat = formatter.format(Instant.EPOCH);
	String epochPlusIntervalFormat;
    Rounding.DateTimeUnit singleUnit = DateHistogramAggregationBuilder.DATE_FIELD_UNITS.get(expression);
	if (singleUnit != null) {
        epochPlusIntervalFormat = formatter.format(Instant.EPOCH.plus(1, singleUnit.getField().getBaseUnit());
    } else {
        // see below
    }

Reply:
I understand that it is a breach of encapsulation. However, `Months` cannot be added to an instant ([see this SO qa] (https://stackoverflow.com/questions/39907925/why-instant-does-not-support-operations-with-chronounit-years)). 

Handling months and quarters is tricky since neither can be reduced to a deterministic set of milliseconds :/. I will try to clean this up while keeping encapsulation.  

I probably should have flagged this PR as WIP as I opened to get somebody else's expertise on java Date handling as I am fairly unfamiliar with that part of the API :)